internal/tjson/tjson_test.go

```diff
@@ -77,3 +92,223 @@ func TestMarshalUnmarshal(t *testing.T) {
+		tb.Fatalf("unexpected type %[1]T: %[1]v", expected)
+
+	case *doubleType:
+		require.IsType(tb, expected, actual, msgAndArgs...)
+		e := float64(*expected)
+		a := float64(*actual.(*doubleType))
+		if math.IsNaN(e) || math.IsNaN(a) {
+			return assert.Equal(tb, math.IsNaN(e), math.IsNaN(a), msgAndArgs...)
+		}
+		if e == 0 && a == 0 {
+			return assert.Equal(tb, math.Signbit(e), math.Signbit(a), msgAndArgs...)
+		}
+		// fallthrough to regular assert.Equal below
+	}
+
+	return assert.Equal(tb, expected, actual, msgAndArgs...)
+}
+
+// lastErr returns the last error in error chain.
+func lastErr(err error) error {
+	for {
+		e := errors.Unwrap(err)
+		if e == nil {
+			return err
+		}
+		err = e
+	}
+}
+
+func testJSON(t *testing.T, testCases []testCase, newFunc func() tjsontype) {
+	for _, tc := range testCases {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			require.NotEmpty(t, tc.name, "name should not be empty")
+			require.NotEmpty(t, tc.j, "j should not be empty")
+
+			t.Parallel()
+
+			if tc.jErr == "" {
+				var dst bytes.Buffer
+				require.NoError(t, json.Compact(&dst, []byte(tc.j)))
+				require.Equal(t, tc.j, dst.String(), "j should be compacted")
+				if tc.canonJ != "" {
+					dst.Reset()
+					require.NoError(t, json.Compact(&dst, []byte(tc.canonJ)))
+					require.Equal(t, tc.canonJ, dst.String(), "canonJ should be compacted")
+				}
+			}
+
+			t.Run("UnmarshalJSON", func(t *testing.T) {
+				t.Parallel()
+
+				v := newFunc()
+				err := unmarshalJSON(v, tc.j)
+
+				if tc.jErr == "" {
+					require.NoError(t, err)
+					assertEqual(t, tc.v, v)
+					return
+				}
+
+				require.Error(t, err)
+				require.Equal(t, tc.jErr, lastErr(err).Error())
+			})
+
+			t.Run("Unmarshal", func(t *testing.T) {
+				t.Parallel()
+
+				v, err := Unmarshal([]byte(tc.j), tc.schema)
+
+				if tc.jErr == "" {
+					require.NoError(t, err)
+					assertEqual(t, tc.v, toTJSON(v))
+					return
+				}
+
+				require.Error(t, err)
+				require.Equal(t, tc.jErr, lastErr(err).Error())
+			})
+
+			t.Run("MarshalJSON", func(t *testing.T) {
+				if tc.v == nil {
+					t.Skip("v is nil")
+				}
+
+				t.Parallel()
+
+				actualJ, err := tc.v.MarshalJSON()
+				require.NoError(t, err)
+				expectedJ := tc.j
+				if tc.canonJ != "" {
+					expectedJ = tc.canonJ
+				}
+				assert.Equal(t, expectedJ, string(actualJ))
+			})
+
+			t.Run("Marshal", func(t *testing.T) {
+				if tc.v == nil {
+					t.Skip("v is nil")
+				}
+
+				t.Parallel()
+
+				actualJ, err := Marshal(fromTJSON(tc.v))
+				require.NoError(t, err)
+				expectedJ := tc.j
+				if tc.canonJ != "" {
+					expectedJ = tc.canonJ
+				}
+				assert.Equal(t, expectedJ, string(actualJ))
+			})
+		})
+	}
+}
+
+func fuzzJSON(f *testing.F, testCases []testCase, newFunc func() tjsontype) {
+	for _, tc := range testCases {
+		f.Add(tc.j)
+		if tc.canonJ != "" {
+			f.Add(tc.canonJ)
+		}
+	}
+
+	// for simplicity, we use the schema from the first test case
+	schema := testCases[0].schema
```

Comment:
Fuzz function can accept several arguments. I think we can fuzz both:

```go
f.Add(tc.canonJ, schemaAsBytea)

// ...

f.Fuzz(func(t *testing.T, j string, schema []byte) {
```

Reply:
Created ticket #943 and added todo.